File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        //정렬 
        Arrays.sort(times);
        long answer = (long)n * times[times.length-1]; //가능한 최댓값 
        
        long left = 0; long right = answer;
        while(left <= right){
            long mid = (left+right)/2;
            
            long cnt = 0;
            for(int t : times){
                cnt += mid/t;
            }

            if(cnt<n){
                left = mid + 1;
            }else{
                right = mid - 1;
                answer = mid;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <climits>

using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    
    long long low = 0;
    sort(times.begin(), times.end());
    
    long long high =  1LL * n * times[times.size()-1];
    
    while(low<=high){
        long long mid = (low+high)/2;
        
        long long cnt = 0;
        for(int t : times){
            cnt+= mid/t;
        }
        
        if(cnt<n){
            low = mid + 1;
        }else{
            high = mid - 1;
            answer = mid;
        }
    }
    
    return answer;
}